embasi/embedding.py

```diff
@@ -906,3 +919,173 @@ def run(self):
+    two levels of theory (low-level and high-level) with the
+    Projection-Based Embedding (PbE) scheme of Manby et al.[1].
+    Parameters
+    ----------
+    atoms: ASE Atoms
+        Structural and chemical information of system
+    embed_mask: int or list
+        Number of atoms from index 0 in layer 1, or list of reigons assigned
+        to each atom (i.e., [1,1,1,2,2,2])
+    calc_base_ll: ASE FileIOCalculator
+        Calculator object for layer 1
+    calc_base_hl: ASE FileIOCalculator
+        Calculator object for layer 2
+    frag_charge: int
+        Charge of the embedded fragment. Defaults to 0.
+    post_scf: str
+        Post-HF method applied to high-level calculation. Defaults to None.
+    mu_val: float
+        Pre-factor for level-shift orthogonalisation. Defaults to 1e+06 Ha.
+    truncate_basis: float or None:
+        Truncates the basis functions of the environment based on
+        a Mulliken charge metrix. Turned off if None. Defaults to None.
+    References
+    ----------
+    [1] Weso
+    """
+
+    def __init__(self, atoms, embed_mask, calc_base_ll, calc_base_hl,
+                 run_dir="./EmbASI_calc"):
+
+        from copy import copy, deepcopy
+        from mpi4py import MPI
+
+        self.calc_names = ["MU0","F2A1","F1A2"]
+
+        super(FrozenDensityEmbedding, self).__init__(atoms,
+                                                     embed_mask,
+                                                     calc_base_ll,
+                                                     calc_base_hl,
+                                                     run_dir=run_dir)
+
+        initial_calculator    = deepcopy(self.calculator_ll)
+        low_level_calculator  = deepcopy(self.calculator_ll)
+        high_level_calculator = deepcopy(self.calculator_hl)
+
+        self.run_dir =  run_dir
+
+        initial_calculator.parameters['qm_embedding_type'] = 'frozendensity  write'
+        low_level_calculator.parameters['qm_embedding_type'] = 'frozendensity  read_and_write'
+        high_level_calculator.parameters['qm_embedding_type'] = 'frozendensity  read_and_write'
+
+        initial_calculator.parameters["aims_output"] = "rho_and_derivs_on_grid"
+        low_level_calculator.parameters['aims_output'] = "rho_and_derivs_on_grid"
+        high_level_calculator.parameters['aims_output'] = "rho_and_derivs_on_grid"
+
+        self.set_layer(atoms, "MU0", initial_calculator, 
+                       embed_mask, ghosts=2, no_scf=False,
+                       insert_embedding_region=False)
+
+        self.set_layer(atoms, "F2A1", low_level_calculator, 
+                       embed_mask, ghosts=2, no_scf=False,
+                       insert_embedding_region=False)
+
+        self.set_layer(atoms, "F1A2", high_level_calculator,
+                       embed_mask, ghosts=1, no_scf=False,
+                       insert_embedding_region=False)
+
+        self.rank = MPI.COMM_WORLD.Get_rank()
```

Comment:
Part of a wider change for myself - we should move this to the master embedding class. This will be useful for all inherited embedding classes, I suspect.